sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/SdkMeterProvider.java

```diff
@@ -56,14 +70,54 @@ public MeterBuilder meterBuilder(@Nullable String instrumentationName) {
     return new SdkMeterBuilder(registry, instrumentationName);
   }
 
-  @Override
-  public Collection<MetricData> collectAllMetrics() {
-    Collection<SdkMeter> meters = registry.getComponents();
-    List<MetricData> result = new ArrayList<>(meters.size());
-    for (SdkMeter meter : meters) {
-      result.addAll(meter.collectAll(sharedState.getClock().now()));
+  /**
+   * Registers a new metric reader on this meter provider.
+   *
+   * @param <R> the type of the reader.
+   * @param factory a constructor for the reader, given access to SDK internals.
+   * @return the registered reader.
+   */
+  public <R extends MetricReader> R register(MetricReader.Factory<R> factory) {
```

Comment:
I would really prefer that we don't allow this mutation to happen after the MeterProvider has been implemented. Why can't this be done up front on the builder?

Reply:
Need to rework how autoconf module works.  It's possible to do this ahead of time, just need to go deeper on refactoring *and* it'll make tests a bit funkier.